modules/dnn/src/int8layers/reduce_layer.cpp

```diff
@@ -0,0 +1,227 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "../precomp.hpp"
+#include "layers_common.hpp"
+
+#include <algorithm>
+#include <stdlib.h>
+#include <numeric>
+
+namespace cv
+{
+namespace dnn
+{
+
+class ReduceLayerInt8Impl CV_FINAL : public ReduceLayerInt8
+{
+public:
+    ReduceLayerInt8Impl(const LayerParams& params)
+    {
+        // Set reduce type
+        CV_Assert(params.has("reduce"));
+        String typeString = toLowerCase(params.get<String>("reduce"));
+        if (typeString == "max")
+            reduceType = MAX;
+        else if (typeString == "min")
+            reduceType = MIN;
+        else
+            CV_Error(Error::StsBadArg, "Unknown reduce type \"" + typeString + "\"");
+
+        // Set deleted dims
+        CV_Assert(params.has("deleted_dims"));
+        DictValue tempDims = params.get("deleted_dims");
+        int i, n = tempDims.size();
+        reduceDims.resize(n);
+        for (i = 0; i < n; i++)
+        {
+            reduceDims[i] = tempDims.get<int>(i);
+        }
+    }
+
+    virtual bool supportBackend(int backendId) CV_OVERRIDE
+    {
+        if (backendId == DNN_BACKEND_OPENCV)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    class ReduceOpBase
+    {
+    public:
+        ReduceOpBase() {}
+        virtual ~ReduceOpBase() {};
+        virtual int8_t apply(const int8_t*, const int8_t*) = 0;
```

Comment:
We still have `virtual` stuff.
It should be dropped as we use templates.

Reply:
Thanks for the code review, I have removed all `virtual` stuff in the base class, both `layers/reduce_layer.cpp` and `int8_layers/reduce_layer.cpp`.